docker/docker.sh

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env bash
+
+set -euxo pipefail
+
+# TODO: Allow this to be configured by metadata.
+readonly DOCKER_VERSION='18.06.0~ce~3-0~debian'
+
+function is_master() {
+  local role="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+  if [[ "$role" == 'Master' ]] ; then
+    true
+  else
+    false
+  fi
+}
+
+function get_docker_gpg() {
+  # Hard code the GPG key to prevent malicious payloads. The key was originally
```

Comment:
Storing a key in a heredoc is pretty unreadable.

In general we trust HTTPS and re-download keys over it on a per installation basis.

If you really don't want to trust docker.com to not be compromised, why not host this key in a GCS bucket we control.

Reply:
Storing it in a GCS bucket seems reasonable, though I do prefer keeping it inline to make the init action more atomic. Why do you find the heredoc unreadable? The function just does one thing and makes that pretty clear.

Alternatively, I _could_ just curl inline here. I didn't want to do that mostly because I'm not sure what happens with `apt-key add` if the payload is something other than a well-formed public key. I don't have a way to verify the key outside of using HTTPS anyway, so all I did here was check by hand that this was in fact the shape we expected.

Reply:
I vote inline then. Curl into | apt-key is pretty standard.